### PR TITLE
For Mobile API, get earned value from CompletionDataMixin instead of StudentProgress model

### PR DESCRIPTION
**Pull request by cgopalan**
The earned value for student progress retrieved from StudentProgress model is not correct. Until this is corrected, get it from the "earned" property of CompletionDataMixin object.

**Testing instructions**:

1. Import a course and complete some problems in the course.
2. Modify the blocks or delete them.
3. Check that the value in the StudentProgress model still shows the original earned completions, but the value is correctly calculated by the backend.

**Author notes and concerns**:

1. This might be a temporary fix.

**Reviewers**
- [ ] @jcdyer
- [ ] @bradenmacdonald

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @cgopalan! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

<a   href="https://openedx-webhooks.herokuapp.com/github/process_pr?number=939&repo=edx-solutions%2Fedx-platform"   rel="nofollow" >Create an OSPR issue for this pull request</a>.

**Comment by cgopalan** [comment]
@bradenmacdonald please review and then merge if you think this works. @jcdyer has already tested it.

**Comment by cgopalan** [comment]
@jcdyer I noticed some tests failed because of the fix. I have fixed the tests. Can you review? Basically the tests that verified that the "earned" value was obtained from the StudentProgress model needed to be changed. This resulted in the earned value being tested for `None` or `0.0` in these cases.
@bradenmacdonald can you trigger a CI build again? The tests should pass now, as I tested it locally.

**Comment by bradenmacdonald** [comment]
@cgopalan Those test changes look wrong to me. The API documentation says that it always returns a float for each field, so we should have any test cases where the API is returning `None` values. In addition, these modified tests are too close to the code - they're now testing implementation details, rather than the desired outcomes. I suggest you leave the test assertions completely unchanged, and instead fix the setup code so that it generates `CourseModuleCompletion` objects (and thus automatically generates `StudentProgress` objects) rather than just generating `StudentProgress` objects directly which then get ignored.

I'll defer to @jcdyer 's recommendations though if he has other thoughts.

File: lms/djangoapps/completion_api/tests/test_serializers.py [hunk 1]
```diff
@@ -91,14 +92,14 @@ def test_serialize_student_progress_object(self, serializer_cls, extra_body):
...
+                'earned': None,
                 'possible': 19.0,
-                'ratio': 16 / 19,
+                'ratio': None,
             },
         }
         expected.update(extra_body)
         self.assertEqual(
-            serial.data,
+            json.loads(json.dumps(serial.data)),
```
Comment (bradenmacdonald) [inline lms/djangoapps/completion_api/tests/test_serializers.py]: There must be a better way to do this than to convert to JSON and back ;)

Comment (cgopalan) [reply lms/djangoapps/completion_api/tests/test_serializers.py]: @bradenmacdonald agreed, I was trying to fix the tests keeping in mind that this is a short-term solution. So to convert a nested OrderedDict to dict, this was the quick fix I came up with.

**Comment by cgopalan** [comment]
@jcdyer I am not familar enough with the tests to fix them the right way as @bradenmacdonald pointed out. Some help would be appreciated.

File: lms/djangoapps/completion_api/tests/test_serializers.py (hunk 1 above)
Comment (jcdyer) [reply lms/djangoapps/completion_api/tests/test_serializers.py]: Was this a problem?  I've never noticed an issue comparing equality between an OrderedDict and an unordered dict.  Or was it more about the readability of the diff?

**Comment by jcdyer** [comment]
@bradenmacdonald 

> so we should have any test cases where the API is returning None values.

I think you meant *shouldn't* here.

@cgopalan 

There are two test classes in this file: One of which checks the formats generated by the serializer using the MockCourseCompletion, and tests some edge cases, and the other which uses the ToyCourseFactory to create a real course, which tests that the hookup between the models and the serializer work appropriately.  

For the first one, you should change the MockCourseCompletion.  Remove its `possible` method, and add a `stats` method that returns appropriate course completion stats.  (See the models.py file for the kind of data that stats should return.

For the test class that works on the ToyCourseFactory, you should do what @bradenmacdonald suggested: Leave the assertions alone, and replace the `StudentProgress.objects.create()` line with a few `CourseModuleCompletion.objects.create()` calls that create completions for video, html, or capa blocks within `self.course`.  (If you grep through the code base for other uses of ToyCourseFactory, you should see how to find blocks within it).  Once you have a block, `block.location` will give you the block_key, aka usage_key, aka usage_id. (Consistent naming is a known issue in the edx-platform code base--nostra culpa).  Then once you've created an appropriate number of `CourseModuleCompletion`s for the test, get the progress object with  `StudentProgress.objects.get(user=whatever, course_id=thinger)`; it will have been auto-generated to match the created `CourseModuleCompletions`.

File: lms/djangoapps/completion_api/tests/test_serializers.py (hunk 1 above)
Comment (cgopalan) [reply lms/djangoapps/completion_api/tests/test_serializers.py]: @jcdyer you are right. That conversion is not necessary. Will remove it.

**Comment by cgopalan** [comment]
@jcdyer I am not sure what kind of data to return in the `stats` method in `MockCourseCompletion`. The model seems to return a stats class, but seems like thats not meant to be used outside of the class. Can you elaborate some more? 
Also, why does the `possible` method need to be removed?

**Comment by jcdyer** [comment]
@cgopalan: Indeed `_Stats` is generally for use inside the CourseCompletionFacade class, but that means it's perfectly appropriate to use in the mock as well!  So if you have MockCourseCompletion.stats return a `_Stats` object, the earned and possible properties will use the _Stats' values to calculate the course-level completion.  

There's no reason you *need* to remove the possible method from the mocked class, but since the regular possible method just pulls its values off the _Stats object, the mocked version becomes redundant.  Leave it if you like.  It's not doing any harm.

**Comment by cgopalan** [comment]
@jcdyer Thanks, I have fixed the serializer tests. However for the `ToyCourseFactory` tests, if I create multiple `CourseCompletion` objects (for video), it doesnt seem to bump up the `earned` value. Also creating a `CourseCompletion` object does not seem to automatically create a `StudentProgress` object. If I do a `StudentProgress.objects.get(user=self.test_user, course_id=self.course.id)` after creating a `CourseCompletion` object, it says `no matching query exists`.
So I had to change the assertion to check for earned value as 1 instead of 3 (since I could do only one block completion).
Any idea why its not bumping up the earned value?

**Comment by jcdyer** [comment]
I'm not sure.  CourseCompletion and CourseModuleCompletion are not the same thing.  CourseModuleCompletion lives in the progress app.  Is that what you're creating?

**Comment by cgopalan** [comment]
@jcdyer sorry I did create `CourseModuleCompletion` objects, FYI.

**Comment by bradenmacdonald** [comment]
@cgopalan I triggered a new test build for you. Do you need help getting this done ASAP?

**Comment by cgopalan** [comment]
@bradenmacdonald the tests should pass. I had to change this assertion line (https://github.com/open-craft/edx-platform/blob/7e7293a735085c1c04e0d0493e2d37b1e52f2743/lms/djangoapps/completion_api/tests/test_serializers.py#L179) from 3 to 1 since I wasnt able to add multiple course completion objects to bump up the earned value to 3. Thats the only thing left.

**Comment by bradenmacdonald** [comment]
@cgopalan It turns out that for the solutions fork's test environment, [there is a flag called `DISABLE_SOLUTIONS_APPS_SIGNALS`](https://github.com/edx-solutions/edx-platform/blob/9a68f82c2d41802cd5b89a494ec1dad0462072e1/lms/envs/test.py#L589) which prevents [the usual signal handler](https://github.com/edx-solutions/progress-edx-platform-extensions/blob/6cf010ffa2c95dce10d14f34b7d3e3b54573ec6b/progress/signals.py#L60) from creating `StudentProgress` objects. To work around this, you need to use the very misleadingly named [`SignalDisconnectTestMixin `](https://github.com/edx-solutions/api-integration/blob/7b118bf5806cebe7f2bdae399f59339e6b33d1f6/edx_solutions_api_integration/test_utils.py#L219), which will re-connect those signals for the duration of the test in question. I've pushed a fix.

@jcdyer Can you please review and give final approval so I can squash and merge ASAP?

**Comment by jcdyer** [comment]
:+1:


- [x] I tested this: Verified that only test code has changed since my last review
- [x] I read through the code.  Tests are now behaving as expected.  
- [x] ~~I checked for accessibility issues~~ N/A (See previous review)
- [x] ~~Includes documentation~~ N/A (See previous review)

**Comment by bradenmacdonald** [comment]
Squashed from [9d5a72aa8d16837f912fafdddc5c50d9b49834ae](https://github.com/open-craft/edx-platform/commits/9d5a72aa8d16837f912fafdddc5c50d9b49834ae) to ee3f5f4ceac2c9824f9cd5da0c2d4b012a6fd214 in preparation for merge.
